subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c

```diff
@@ -132,6 +138,87 @@ static void azure_iot_hub_notify_event(struct azure_iot_hub_evt *evt)
 	}
 }
 
+const char *state_name_get(enum connection_state state)
+{
+	switch (state) {
+	case STATE_IDLE: return "STATE_IDLE";
+	case STATE_INIT: return "STATE_INIT";
+	case STATE_CONNECTING: return "STATE_CONNECTING";
+	case STATE_CONNECTED: return "STATE_CONNECTED";
+	case STATE_DISCONNECTING: return "STATE_DISCONNECTING";
+	default: return "STATE_UNKNOWN";
+	}
+}
+
+static void connection_state_set(enum connection_state new_state)
+{
+	bool notify_error = false;
+
+	/* Check for legal state transitions. */
+	switch (connection_state) {
+	case STATE_IDLE:
+		if (new_state != STATE_INIT) {
+			notify_error = true;
+		}
+		break;
+	case STATE_INIT:
+		if (!(new_state != STATE_CONNECTING ||
+		      new_state != STATE_INIT)) {
+			notify_error = true;
+		}
+		break;
+	case STATE_CONNECTING:
+		if (!(new_state != STATE_CONNECTED ||
+		      new_state != STATE_INIT)) {
+			notify_error = true;
+		}
+		break;
+	case STATE_CONNECTED:
+		if (!(new_state != STATE_DISCONNECTING ||
+		      new_state != STATE_INIT)) {
+			notify_error = true;
+		}
+		break;
+	case STATE_DISCONNECTING:
+		if (new_state != STATE_INIT) {
+			notify_error = true;
+		}
+		break;
+	default:
+		LOG_ERR("New connection state unknown");
+		notify_error = true;
+		break;
+	}
+
+	if (notify_error) {
+		struct azure_iot_hub_evt evt = {
+			.type = AZURE_IOT_HUB_EVT_ERROR,
+			.data.err = -EINVAL
+		};
+
+		LOG_ERR("Invalid connection state transition, %s --> %s",
+			log_strdup(state_name_get(connection_state)),
+			log_strdup(state_name_get(new_state)));
+
+		azure_iot_hub_notify_event(&evt);
+		return;
+	}
+
+	connection_state = new_state;
+
+	LOG_DBG("New connection state: %s",
+		log_strdup(state_name_get(connection_state)));
+}
+
+static bool connection_state_verify(enum connection_state state)
+{
+	if (connection_state == state) {
```

Comment:
Up to preference really, but this could be as simple as `return (connection_state == state);`. No strong push though if you prefer the current way.

Reply:
Sure. Why not. Lets save some extra lines.